jax/scipy/stats/multivariate_normal.py

```diff
@@ -0,0 +1,49 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as onp
+import scipy.stats as osp_stats
+
+from ... import lax
+from ...numpy.lax_numpy import _promote_args_like, _constant_like, _wraps
+from ...numpy.lax_numpy import dot, subtract, einsum
+from ...numpy.linalg import det, inv
+
+
+@_wraps(osp_stats.multivariate_normal.logpdf)
+def logpdf(x, mean, cov):
+  # x, mean, cov = _promote_args_like(osp_stats.multivariate_normal.logpdf, x, mean, cov)
```

Comment:
Is this line not needed anymore? If not, maybe just delete it instead of leaving it commented out.

But it seems more likely that we want promotion logic here, rather than casting everything to cov.dtype. Can we just put this line back in (replacing the `astype` lines immediately below), or is there a problem with it?

Reply:
I commented that line out because it led to the following error. I think it's because the positive-definite check in `scipy.stats.multivariate_normal` brings some problem to `_promote_args_like`. I didn't know how to fix this so I used `astype` instead. Could you give me some ideas on this?

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-437963fd1205> in <module>()
----> 1 scipy.stats.multivariate_normal.pdf(np.array([1,2]),np.array([3,4]),np.eye(2))

/global/u2/h/hejia/jax/jax/scipy/stats/multivariate_normal.py in pdf(x, mean, cov)
     47 @_wraps(osp_stats.multivariate_normal.pdf)
     48 def pdf(x, mean, cov):
---> 49   return lax.exp(logpdf(x, mean, cov))

/global/u2/h/hejia/jax/jax/scipy/stats/multivariate_normal.py in logpdf(x, mean, cov)
     28 @_wraps(osp_stats.multivariate_normal.logpdf)
     29 def logpdf(x, mean, cov):
---> 30   x, mean, cov = _promote_args_like(osp_stats.multivariate_normal.logpdf, x, mean, cov)
     31   x = x.astype(cov.dtype)
     32   mean = mean.astype(cov.dtype)

/global/u2/h/hejia/jax/jax/numpy/lax_numpy.py in _promote_args_like(op, *args)
    195   """Convenience function to apply shape and dtype promotion to result type."""
    196   _check_arraylike(op.__name__, *args)
--> 197   return _promote_shapes(*_promote_to_result_dtype(op, *args))
    198 
    199 

/global/u2/h/hejia/jax/jax/numpy/lax_numpy.py in _promote_to_result_dtype(op, *args)
    167 def _promote_to_result_dtype(op, *args):
    168   """Convenience function to promote args directly to the op's result dtype."""
--> 169   to_dtype = _result_dtype(op, *args)
    170   return [lax.convert_element_type(arg, to_dtype) for arg in args]
    171 

/global/u2/h/hejia/jax/jax/numpy/lax_numpy.py in _result_dtype(op, *args)
    174   """Compute result dtype of applying op to arguments with given dtypes."""
    175   args = (onp.ones((0,) * ndim(arg), _dtype(arg)) for arg in args)
--> 176   return _dtype(op(*args))
    177 
    178 

~/.conda/envs/hejia@cori/lib/python3.6/site-packages/scipy/stats/_multivariate.py in logpdf(self, x, mean, cov, allow_singular)
    485         dim, mean, cov = self._process_parameters(None, mean, cov)
    486         x = self._process_quantiles(x, dim)
--> 487         psd = _PSD(cov, allow_singular=allow_singular)
    488         out = self._logpdf(x, mean, psd.U, psd.log_pdet, psd.rank)
    489         return _squeeze_output(out)

~/.conda/envs/hejia@cori/lib/python3.6/site-packages/scipy/stats/_multivariate.py in __init__(self, M, cond, rcond, lower, check_finite, allow_singular)
    150         # Note that eigh takes care of array conversion, chkfinite,
    151         # and assertion that the matrix is square.
--> 152         s, u = scipy.linalg.eigh(M, lower=lower, check_finite=check_finite)
    153 
    154         eps = _eigvalsh_to_eps(s, cond, rcond)

~/.conda/envs/hejia@cori/lib/python3.6/site-packages/scipy/linalg/decomp.py in eigh(a, b, lower, eigvals_only, overwrite_a, overwrite_b, turbo, eigvals, type, check_finite)
    430         if eigvals is None:
    431             w, v, info = evr(a1, uplo=uplo, jobz=_job, range="A", il=1,
--> 432                              iu=a1.shape[0], overwrite_a=overwrite_a)
    433         else:
    434             (lo, hi) = eigvals

ValueError: failed to create intent(cache|hide)|optional array-- must have defined dimensions but got (0,)
```

Reply:
Sorry for being slow to respond on this one.

I see, thanks for explaining. Could you add a short comment to the code with this information? Maybe something like:

```
# TODO add back dtype promotion after resolving the PSD-check issue
# see https://github.com/google/jax/pull/268
```

Reply:
On second thought, I can add that TODO note myself after we merge. No need to slow this down :)